Align ref and weakRef counting on uint32_t

annevk opened the pull request:
#### 26d2fd2eb3613de67453dbb8e74315831d7b85bb
<pre>
Align ref and weakRef counting on uint32_t
<a href="https://bugs.webkit.org/show_bug.cgi?id=304564">https://bugs.webkit.org/show_bug.cgi?id=304564</a>

Reviewed by Simon Fraser.

uint32_t is more explicit than unsigned and can reduce memory compared
to size_t on 64-bit platforms. And given that only ThreadSafeWeakPtr is
using size_t at the moment, this should be a reasonable change to make.

Canonical link: <a href="https://commits.webkit.org/304921@main">https://commits.webkit.org/304921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70a5d1253f14b4cba976ea7bba9e57120c898c52

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136688 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9047 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47975 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144416 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89661 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4aa41d8b-42ae-4b1a-b3b0-8b2bda43c392) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9748 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8893 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104529 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7d95fa29-9a36-405b-ae98-bd6f37187b5f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7125 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122479 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85368 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/67003a03-d0b0-48b1-b97d-c69d2d6fa08c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6769 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4456 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5008 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128649 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116088 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40667 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147173 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135174 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8731 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41240 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112881 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8749 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7348 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113210 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6693 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118770 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62865 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21100 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8779 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36824 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167954 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8500 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72345 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43822 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8719 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8571 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1113,7 +1113,7 @@ inline Ref<StringImpl> StringImpl::adopt(Vector<CharacterType, inlineCapacity, O
 }
 
-inline size_t StringImpl::cost() const
+inline uint32_t StringImpl::cost() const
```
This doesn't look right. This is not strictly related to ref-count. This is a memory cost and restricting that to uint32_t is not great.

annevk replied:
Can it exceed it though given that m_length is unsigned? The problem is that divideRoundedUp does not work otherwise.

An alternative is that I exclude StringImpl from this patch.

cdumez replied:
It was doing `size_t result = m_length;` Then `result <<= 1;`, so yes, I think the value could exceed that uint32_max before in theory.

cdumez replied:
for `divideRoundedUp`, you can probably call `divideRoundedUp<size>()` instead of `divideRoundedUp()` to help it figure out the best common type.

cdumez commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1127,21 +1127,21 @@ inline size_t StringImpl::cost() const
 }
 
-inline size_t StringImpl::costDuringGC()
+inline uint32_t StringImpl::costDuringGC()
```
ditto.

smfr commented on Source/WebCore/Modules/webdatabase/DatabaseThread.cpp:
```diff
@@ -105,7 +105,7 @@ void DatabaseThread::databaseThread()
     m_transactionCoordinator->shutdown();
 
-    LOG(StorageAPI, "About to detach thread %p and clear the ref to DatabaseThread %p, which currently has %zu ref(s)", m_thread.get(), this, refCount());
+    LOG(StorageAPI, "About to detach thread %p and clear the ref to DatabaseThread %p, which currently has %i ref(s)", m_thread.get(), this, refCount());
```
LOG_WITH_STREAM would avoid the need for the type specifiers.

annevk committed:
Align ref and weakRef counting on uint32_t
https://bugs.webkit.org/show_bug.cgi?id=304564

Reviewed by Simon Fraser.

uint32_t is more explicit than unsigned and can reduce memory compared
to size_t on 64-bit platforms. And given that only ThreadSafeWeakPtr is
using size_t at the moment, this should be a reasonable change to make.

Canonical link: https://commits.webkit.org/304921@main
